### PR TITLE
Made Beam class accept cross-section geometries

### PR DESCRIPTION
**Pull request by ishanaj**
As discussed in PR #16964 , the beam class is made to accept geometry object as an option instead of the second moment.
Instead of the `second moment`, beam class now accepts a `shape` as an argument.
This argument `shape` is the description to the cross-section of the beam.
It is either the shape of the cross section (Geometry object) or the second moment of the beam cross-section.
This introduction of geometries is limited to 2D beams. We can discuss it for 3d beams further

TODO's:
- [x] Adding Tests
- [ ] ~~Beam3D~~

#### Release Notes

- physics.continuum_mechanics
    - Beam class now accepts Geometry object representing the shape of the cross-section

### CODE REVIEW

**Comment by ishanaj** [comment]
ping @jashan498 @moorepants @oscarbenjamin

File: sympy/physics/continuum_mechanics/beam.py
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: Don't change the argument name in case people called it explicitly. `second_moment=I`. This retains backwards compatibly.

File: sympy/physics/continuum_mechanics/beam.py
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: You need to retain the prior explanation and add the additional describing the shape.

File: sympy/physics/continuum_mechanics/beam.py
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: Don't change this for backwards compatibility.

File: sympy/physics/continuum_mechanics/beam.py
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: Making second_moment a list of three items is backwards incompatible.

File: sympy/physics/continuum_mechanics/beam.py
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: Backwards incompatible change.

File: sympy/physics/continuum_mechanics/tests/test_beam.py
Comment (moorepants) [inline sympy/physics/continuum_mechanics/tests/test_beam.py]: These tests should note be changed (backwards incompatible).

**Comment by moorepants** [comment]
I think you are on the right track. Adding a new attribute is a good idea. I'd recommend calling the new attribute `cross_section`. But you need to ensure backwards compatibility. I recommend not changing any existing tests, but writing the tests for the new desired functionality first. Then try ot change the class to get it to pass both set of tests. Writing the tests first will force you to think about the desired outcome more carefully.

**Comment by ishanaj** [comment]
Sure, will work on it.

File: sympy/physics/continuum_mechanics/beam.py
Comment (oscarbenjamin) [inline sympy/physics/continuum_mechanics/beam.py]: Sympifyable or Sympifiable? Not sure...

File: sympy/physics/continuum_mechanics/beam.py
Comment (oscarbenjamin) [inline sympy/physics/continuum_mechanics/beam.py]: I would keep the above as it was and instead of the last sentence write:
```
Alternatively ``cross_section`` can be a shape object such as a ``Polygon``
from the geometry module. The second moment of area will be computed
from the shape object.
```
This should either be all one paragraph (no line break) or should have a blank line to separate the paragraphs.

File: sympy/physics/continuum_mechanics/beam.py
Comment (oscarbenjamin) [inline sympy/physics/continuum_mechanics/beam.py]: I think this logic should go in `__init__` rather than `second_moment.setter`.

File: sympy/physics/continuum_mechanics/beam.py
Comment (oscarbenjamin) [inline sympy/physics/continuum_mechanics/beam.py]: Should this not update second_moment?

Comment (ishanaj) [reply sympy/physics/continuum_mechanics/beam.py]: It is being updated in the second moment setter

File: sympy/physics/continuum_mechanics/tests/test_beam.py
Comment (oscarbenjamin) [inline sympy/physics/continuum_mechanics/tests/test_beam.py]: Are the lines above needed? I thought the whole point was that we can pass a shape argument when constructing the beam and then the second moment would be automatically calculated.

Also is it necessary in the API to be able to modify the cross_section after creating the beam (seems like poor API to me)? If it is necessary then at least it should be enough to set the cross_section and have the second_moment be calculated implicitly from that.

File: sympy/physics/continuum_mechanics/beam.py
Comment (ishanaj) [reply sympy/physics/continuum_mechanics/beam.py]: I will correct it!

File: sympy/physics/continuum_mechanics/beam.py
Comment (ishanaj) [reply sympy/physics/continuum_mechanics/beam.py]: I guess then it would have an impact on the `Beam3D` class, as it is not linked with geometries.
Since the second moment setter is overridden in `Beam3D`, it won't create a problem.

File: sympy/physics/continuum_mechanics/tests/test_beam.py
Comment (ishanaj) [reply sympy/physics/continuum_mechanics/tests/test_beam.py]: It is being calculated implicitly, these two lines are just testing whether correct values has been calculated (using assert).

Comment (ishanaj) [reply sympy/physics/continuum_mechanics/tests/test_beam.py]: But I think it won't be able to update it if the cross_section is explicitly changed later. We should better make `cross_section` read-only.

Comment (oscarbenjamin) [reply sympy/physics/continuum_mechanics/tests/test_beam.py]: Sorry I misread the lines above as changing the properties not checking them. That seems fine.

In general I don't see the benefit in being able to change the geometry after the object is created but beam seems to have a bunch of things like this (`add_load` etc).

File: sympy/physics/continuum_mechanics/beam.py
Comment (oscarbenjamin) [reply sympy/physics/continuum_mechanics/beam.py]: So if I create a Beam with a particular second_moment and then change the geometry with `beam.cross_section = new_shape` will the `_second_moment` attribute be updated?

File: sympy/physics/continuum_mechanics/tests/test_beam.py
Comment (moorepants) [reply sympy/physics/continuum_mechanics/tests/test_beam.py]: > Also is it necessary in the API to be able to modify the cross_section after creating the beam (seems like poor API to me)? 

I'd be fine with making cross section "private" attribute, then we don't need to support people changing it after creating a beam. You can also not have a setter, to make it impossible to modify.

The Beam class is a mutable object for better or worse. It lets you construct it piecemeal instead of all in one go. This is likely an advantage for complex beams and for interactive work. Maybe we should stick with following its existing design patterns.

Comment (ishanaj) [reply sympy/physics/continuum_mechanics/tests/test_beam.py]: Should we make cross-section read-only instead of private, by removing setter but retaining property? 
Also does the `if statement` has to be shifted to `__init__` or is it fine there? I think with the setter being removed it should have no problem with the Beam3D class.

Comment (jashan498) [reply sympy/physics/continuum_mechanics/tests/test_beam.py]: Yeah, we are making `_cross_section` read-only. You have to remove the `cross_section` setter for that. We don't want the user to change `_cross_section` attribute using it. As we are making it read-only, I think property method should be there.

File: sympy/physics/continuum_mechanics/beam.py
Comment (jashan498) [inline sympy/physics/continuum_mechanics/beam.py]: I just realized that putting this `if` block inside `init` will make it run only once (during the creation of the object). Therefore, making changes to `second_moment` won't change `_cross_section`.
```
>>> b1 = Beam(50, E, Circle((0, 0), r))
>>> b1.cross_section
Circle(Point2D(0, 0), r)
>>> b1.second_moment
pi*r*Abs(r)**3/4
>>> b1.second_moment = Polygon((0, 0), (a, 0), (a, c), (0, c))
>>> b1.cross_section
Circle(Point2D(0, 0), r)   # should be Polygon((0, 0), (a, 0), (a, c), (0, c))
>>> b1.second_moment
Polygon(Point2D(0, 0), Point2D(a, 0), Point2D(a, c), Point2D(0, c))  # should be a*c**3/12

```

Comment (ishanaj) [reply sympy/physics/continuum_mechanics/beam.py]: >  `>>>`b1.second_moment = Polygon((0, 0), (a, 0), (a, c), (0, c))
      b1.cross_section

But do we want this kind of thing to be supported? If so then it would better be with the `cross_section` attribute being changed, like:
``` 
b1.cross_section= Polygon((0, 0), (a, 0), (a, c), (0, c)) 
```
And for this the setter will have to be introduced.

Comment (ishanaj) [reply sympy/physics/continuum_mechanics/beam.py]: And for that we can just add this setter in the current code, making if condition to be in `__init__` where it is ideally supposed to be. But `cross_section` won't be private then
```
@cross_section.setter
    def cross_section(self, s):
        if s:
            self.second_moment = s.second_moment_of_area()[0]
        self._cross_section = s
```

Comment (jashan498) [reply sympy/physics/continuum_mechanics/beam.py]: > But do we want this kind of thing to be supported?

As long as there is a setter method for `second_moment`, the user should be able to change the property using that setter. 
Saying that we can remove the setter for `second_moment` too and make it read-only but then it won't be consistent with other similar attributes of Beam (`length`, `elastic_modulus` & `variable`).

Comment (ishanaj) [reply sympy/physics/continuum_mechanics/beam.py]: Is there any harm in making the `cross_section` attribute non-private, with a setter?
If there is no harm, then I guess we can go with the setter that I have mentioned above.
It would also help in updating the second moment when the cross-section is updated by the user.

Comment (moorepants) [reply sympy/physics/continuum_mechanics/beam.py]: If you allow second_moment and cross_section to be mutable you need to make sure a user can add both after creation and that if they add one the other is removed.

Comment (ishanaj) [reply sympy/physics/continuum_mechanics/beam.py]: > make sure a user can't add both simultaneously after creation and that if they add one the other is removed.

I have added tests for it. Please have a look

File: sympy/physics/continuum_mechanics/beam.py
Comment (smichr) [inline sympy/physics/continuum_mechanics/beam.py]:
```suggestion
            of area will be computed from the Polygon internally.
```

File: sympy/physics/continuum_mechanics/beam.py
Comment (smichr) [inline sympy/physics/continuum_mechanics/beam.py]:
```suggestion
            representing the Beam's second moment of area. It is a geometrical
```

File: sympy/physics/continuum_mechanics/beam.py
Comment (smichr) [inline sympy/physics/continuum_mechanics/beam.py]:
```suggestion
            Describes the cross-section of the beam via a SymPy expression
```

File: sympy/physics/continuum_mechanics/beam.py
Comment (ishanaj) [reply sympy/physics/continuum_mechanics/beam.py]: But ellipses, and circles are also being considered, so is it okay to specify 'Polygon' here?

**Comment by ishanaj** [comment]
For cases where the user tries to add a geometry object to the `second_moment` attribute, I think we should raise a `ValueError` with a message redirecting it to change the geometry object.

```
>>> b.second_moment = Polygon((0, 0), (a, 0), (a, b), (0, b))
ValuError: To update cross-section geometry use `cross_section` attribute
```

File: sympy/physics/continuum_mechanics/beam.py
Comment (oscarbenjamin) [reply sympy/physics/continuum_mechanics/beam.py]: Polygon is too specific but there isn't a good term in the mro to describe what you want to accept:
```
In [22]: Polygon.mro()                                                                                                            
Out[22]: 
[sympy.geometry.polygon.Polygon,
 sympy.geometry.entity.GeometrySet,
 sympy.geometry.entity.GeometryEntity,
 sympy.sets.sets.Set,
 sympy.core.basic.Basic,
 object]
```
I would just use "shape" in the docstring. In any case "gemetry" is missing an "o".

**Review by jashan498 — APPROVED** [review]
Looks good to me, can be merged once it gets approved by others too.

**Comment by oscarbenjamin** [comment]
Looks fine to me

File: sympy/physics/continuum_mechanics/beam.py
Comment (moorepants) [inline sympy/physics/continuum_mechanics/beam.py]: Being that this selects the Ixx value, I think the documentation should reflect this. Say something like "If a Geometry object is used, it is assumed that the X axis of the object is aligned with the bending axis.".

**Comment by moorepants** [comment]
Looks good to me except for the one comment I left about being explicit in the docs about whether it selects Ixx or Iyy. Very nice addition!

File: sympy/physics/continuum_mechanics/beam.py
Comment (ishanaj) [reply sympy/physics/continuum_mechanics/beam.py]: I have made the change.  Please have a look

Comment (moorepants) [reply sympy/physics/continuum_mechanics/beam.py]: Looks good. It can be merged with the tests pass.

**Comment by ishanaj** [comment]
The checks have passed. I guess we are good to go!

**Comment by moorepants** [comment]
Nice work! Thanks for adjusting to all the feedback.

**Comment by ishanaj** [comment]
Thanks!

**Comment by moorepants** [comment]
As a companion to this, it would be nice to see more examples in the Sphinx docs showing off this new feature.

**Pull request merged**
